### ISSUE TITLE
Debug info?

### ISSUE THREAD

**Issue by jaydiablo** [geckoboard/sql-dataset#21]
I was testing this tool out today and ran into this error message when trying to append some data to a dataset:

>There was an error while trying to update dataset.name: Sorry, there seems to be a problem with Geckoboard's servers. Please try again, or checkhttps://geckoboard.statuspage.io

Looking at the code, this seems to happen when there's an unexcepted http status code (not 200 - 300 or 400 - 500).  Unfortunately I don't know where to go from here since there doesn't seem to be a good way to get any more information about the error.

Is there a flag or something I can pass to the binary to get more verbose information output to the console?  Perhaps the request and response?

**Comment by jaydiablo** [comment]
I found the debug build attached to #17 and tried running with that, unfortunately it doesn't provide much more information:

>DEBUG: Making request to Geckoboard
DEBUG: Finished making request to Geckoboard
Dataset 'dataset.name' errored: Unexpected server error response from Geckoboard

This is after it makes a query to the database and scans the row values.

**Comment by jaydiablo** [comment]
In case it's useful, the debug build is this version:

>Version: 0.2.1
GitSHA: a9797ee

The non-debug is this:

>Version: 0.2.2
GitSHA: ab69898

**Comment by jnormington** [comment]
Hey @jaydiablo 

Sorry your having issues with sql-dataset and thank you for the information. Unfortunately we don't have a debug switch but something I'd like to add at some point. 

It seems based on the debug error message you are getting a 50x status code as you've noted. I'm not sure it will return any useful information in the body but you can find attached a debug version. Which will print the request body, uri and the response status code and body for 50x case.

Hopefully that will help diagnose the issue

[debug-builds.zip](https://github.com/geckoboard/sql-dataset/files/2125133/builds.zip)

**Comment by jaydiablo** [comment]
Thanks for the quick reply and the debug build.  Unfortunately, as you suggested, not much information is available in the response:

>DEBUG: RequestURL - "https://api.geckoboard.com/datasets/dataset.name/data"
DEBUG: StatusCode - 500
DEBUG: Response - An unhandled lowlevel error occurred. The application logs may have details.

I'll try some different things with the SQL and see if I can get something to post successfully.

**Comment by jnormington** [comment]
Hey @jaydiablo,

Interesting error, if you send a ticket into support with your details - we can investigate further to understand what error on our side is causing that issue.

With the support ticket it would be good to supply the dataset name you are pushing too as well as the payload from the debug output so that we can investigate.

Our great CS team might see something in the logs or a problem with the payload sql-dataset has generated

**Comment by jaydiablo** [comment]
I've been able to get it to successfully POST by reducing the number of rows I'm selecting (I was selecting 500, and reduced it to 100 for this test).

Perhaps the payload with 500 rows is too large.

I'll look into submitting a support ticket, thanks for your help!

**Comment by jnormington** [comment]
Reason I ask what OS +arch saves me waiting for all versions to build :)

**Comment by jnormington** [comment]
@jaydiablo new binaries so as you already say you was appending this should work fine (it will break replace type datasets in a way saying that you can only send x) the number you specify.

but if you pass a new switch `--maxRows 100` try 100 to begin with and see how that goes if it doesn't error - I guess it could well be something with the body length - if it does error try to reduce it further to identify  the batch it fails with.

[builds.zip](https://github.com/geckoboard/sql-dataset/files/2125306/builds.zip)

**Comment by jaydiablo** [comment]
I'm using the Linux amd64 binary.

I'll play with that new build in a bit, thanks!

**Comment by jnormington** [comment]
Great @jaydiablo just to clarify so your SQL is selecting the full rowset as you originally was and letting sql-dataset send them in batches to Geckoboard of your custom input works just fine ?

Yeah would seem like that... Thanks for checking that and feeding back your findings

**Comment by jnormington** [comment]
Hi @jaydiablo,

Just wondered if you'd had a chance to try the latest build and what the outcome was ?

**Comment by jaydiablo** [comment]
@jnormington yes, I did get to try that build finally, and this is what I've found so far.

Even with the batch set to 100, I still received a failure response from the geckoboard servers (same one as before), but only on the last batch (the first 4 batches posted fine).

I tried looking through the JSON body of the last batch for special characters or something that might be causing the issue...  I found a couple of characters that were being converted to escaped UTF sequences (like ampersand and a preceding whitespace character). As a test, I removed or converted those characters in the database and/or the SQL statement, to no avail.

That's as far as I've been able to test so far, I should be able to spend more time on it later today.
